### PR TITLE
Fix params

### LINKED ISSUE #637: nccs mixing implicit and explicit slot index in debug info
You can't mix and match implicit and explicit slot indexes in debug info. Either all the variables in a given array must all be implicitly or explicitly indexed.

Example from the registrar sample:

``` json
{
    "id": "DevHawk.Contracts.DomainStorage.DomainStorage(string)",
    "name": "DevHawk.Contracts.DomainStorage,.ctor",
    "range": "212-230",
    "params": [
        "this,Any", // this param has no slot info
        "prefix,String,0" // this param has slot info
    ],
    "return": "Void",
    "variables": [],
    "sequence-points": [
        "215[0]114:9-114:10",
        "216[0]115:13-115:73",
        "230[0]116:9-116:10"
    ]
},
```

### PR DESCRIPTION
**Pull request by erikzhang**
Fix #637

### CODE REVIEW

**Review by shargon — DISMISSED** [review]
@devhawk could you check if it's solve your issue?

File: src/Neo.Compiler.CSharp/CompilationContext.cs [hunk 1]
```diff
@@ -310,8 +310,9 @@ public JObject CreateDebugInformation()
                     ["id"] = m.Symbol.ToString(),
                     ["name"] = $"{m.Symbol.ContainingType},{m.Symbol.Name}",
                     ["range"] = $"{m.Instructions[0].Offset}-{m.Instructions[^1].Offset}",
-                    ["params"] = (m.Symbol.IsStatic ? Array.Empty<JString>() : new JString[] { "this,Any" })
-                        .Concat(m.Symbol.Parameters.Select((p, i) => (JString)$"{p.Name},{p.Type.GetContractParameterType()},{i}"))
+                    ["params"] = (m.Symbol.IsStatic ? Array.Empty<string>() : new string[] { "this,Any" })
+                        .Concat(m.Symbol.Parameters.Select(p => $"{p.Name},{p.Type.GetContractParameterType()}"))
+                        .Select((p, i) => (JString)$"{p},{i}")
```
Comment (devhawk) [inline src/Neo.Compiler.CSharp/CompilationContext.cs:315]: This looks fine.

**Comment by devhawk** [comment]
cc @erikzhang @shargon 

Still waiting on this fix. Debugging instance methods is broken in RC3 because of this. A method with this signature:

``` csharp
public ByteString MintNFT(string name, string description, string image)
```

generates debug info that looks like this:

``` json
"params": [
    "this,Any",
    "name,String,0",
    "description,String,1",
    "image,String,2"
],
```

Not only is the `this,Any` value missing a slot index, the slot indexes of the other params are all off by one. The debug info *should* look like this: 


``` json
"params": [
    "this,Any,0",
    "name,String,1",
    "description,String,2",
    "image,String,3"
],
```

**Comment by shargon** [comment]
@devhawk could you review it again?

**Comment by erikzhang** [comment]
@neo-project/ngd-shanghai Please test it.

**Comment by superboyiii** [comment]
>@neo-project/ngd-shanghai Please test it.

Looks good on mine. The slot indexes of the other params are all off by one.
```
{
    "hash":"0x437bd785e516d88dacee0a4c9b0da58bf458ddd5",
    "documents":[
        "E:\\Release\\RC4\\TestNetFee\\TestNetFee.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/ByteString.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/CallingConventionAttribute.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/ContractAttribute.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/ContractHashAttribute.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/ContractParameterType.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/ContractPermissionAttribute.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/ContractTrustAttribute.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/ECPoint.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/ExecutionEngine.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Helper.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/IApiInterface.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/InitialValueAttribute.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/List.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/ManifestExtraAttribute.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Map.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Native/ContractManagement.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Native/CryptoLib.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Native/GAS.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Native/Ledger.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Native/NamedCurve.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Native/NEO.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Native/Oracle.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Native/OracleResponseCode.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Native/Policy.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Native/Role.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Native/RoleManagement.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Native/StdLib.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Nep11Token.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Nep11TokenState.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Nep17Token.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/OpCode.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/OpCodeAttribute.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Properties/AssemblyInfo.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/SafeAttribute.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Services/Block.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Services/CallFlags.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Services/Contract.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Services/Crypto.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Services/FindOptions.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Services/Iterator.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Services/Notification.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Services/Runtime.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Services/Storage.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Services/StorageContext.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Services/StorageMap.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Services/Transaction.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/Services/TriggerType.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/SmartContract.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/StackItemType.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/SupportedStandardsAttribute.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/SyscallAttribute.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/TokenContract.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/UInt160.cs",
        "C:\\Users\\SPIRIT\\.nuget\\packages\
eo.smartcontract.framework/3.0.0-rc2\\src/UInt256.cs"
    ],
    "static-variables":[
        "Owner,Hash160,0",
        "assetMap,Any,1",
        "ProxyScriptHash,ByteArray,2"
    ],
    "methods":[
        {
            "id":"TestNetFee.TestNetFee.IsOwner()",
            "name":"TestNetFee.TestNetFee,IsOwner",
            "range":"0-6",
            "params":[

            ],
            "return":"Boolean",
            "variables":[

            ],
            "sequence-points":[
                "0[0]28:42-28:69"
            ]
        },
        {
            "id":"TestNetFee.TestNetFee.Verify()",
            "name":"TestNetFee.TestNetFee,Verify",
            "range":"13-15",
            "params":[

            ],
            "return":"Boolean",
            "variables":[

            ],
            "sequence-points":[
                "13[0]33:40-33:49"
            ]
        },
        {
            "id":"TestNetFee.TestNetFee.GetPaymentStatus()",
            "name":"TestNetFee.TestNetFee,GetPaymentStatus",
            "range":"16-48",
            "params":[

            ],
            "return":"Boolean",
            "variables":[

            ],
            "sequence-points":[
                "16[0]38:50-38:96"
            ]
        },
        {
            "id":"TestNetFee.TestNetFee.EnablePayment()",
            "name":"TestNetFee.TestNetFee,EnablePayment",
            "range":"72-121",
            "params":[

            ],
            "return":"Void",
            "variables":[

            ],
            "sequence-points":[
                "72[0]43:17-43:27",
                "77[0]43:29-43:70",
                "97[0]44:13-44:39",
                "121[0]45:9-45:10"
            ]
        },
        {
            "id":"TestNetFee.TestNetFee.DisablePayment()",
            "name":"TestNetFee.TestNetFee,DisablePayment",
            "range":"137-189",
            "params":[

            ],
            "return":"Void",
            "variables":[

            ],
            "sequence-points":[
                "137[0]50:17-50:27",
                "145[0]50:29-50:70",
                "165[0]51:13-51:39",
                "189[0]52:9-52:10"
            ]
        },
        {
            "id":"TestNetFee.TestNetFee.OnNEP17Payment(Neo.UInt160, System.Numerics.BigInteger, object)",
            "name":"TestNetFee.TestNetFee,OnNEP17Payment",
            "range":"190-243",
            "params":[
                "from,Hash160,0",
                "amount,Integer,1",
                "data,Any,2"
            ],
            "return":"Void",
            "variables":[

            ],
            "sequence-points":[
                "193[0]57:17-57:35",
                "200[0]59:17-59:24",
                "204[0]63:17-63:77",
                "243[0]65:9-65:10"
            ]
        },
        {
            "id":"TestNetFee.TestNetFee.Unlock(Neo.UInt160, Neo.UInt160, System.Numerics.BigInteger)",
            "name":"TestNetFee.TestNetFee,Unlock",
            "range":"244-338",
            "params":[
                "toAssetHash,Hash160,0",
                "toAddress,Hash160,1",
                "amount,Integer,2"
            ],
            "return":"Boolean",
            "variables":[
                "Params,Array,0"
            ],
            "sequence-points":[
                "247[0]69:17-69:27",
                "255[0]69:29-69:70",
                "275[0]70:17-70:95",
                "286[0]72:13-72:75",
                "305[0]73:13-73:57",
                "333[0]74:13-74:25",
                "338[0]75:9-75:10"
            ]
        },
        {
            "id":"TestNetFee.TestNetFee.IsPaused()",
            "name":"TestNetFee.TestNetFee,IsPaused",
            "range":"351-416",
            "params":[

            ],
            "return":"Boolean",
            "variables":[
                "Params,Array,0",
                "isPaused,Boolean,1"
            ],
            "sequence-points":[
                "354[0]79:17-79:42",
                "357[0]81:18-81:109",
                "389[0]82:13-82:37",
                "413[0]83:13-83:29",
                "416[0]84:9-84:10"
            ]
        },
        {
            "id":"TestNetFee.TestNetFee._deploy(object, bool)",
            "name":"TestNetFee.TestNetFee,_deploy",
            "range":"431-463",
            "params":[
                "data,Any,0",
                "update,Boolean,1"
            ],
            "return":"Void",
            "variables":[

            ],
            "sequence-points":[
                "434[0]88:17-88:23",
                "437[0]88:25-88:32",
                "439[0]91:13-91:67",
                "463[0]92:9-92:10"
            ]
        },
        {
            "id":"TestNetFee.TestNetFee.Update(Neo.SmartContract.Framework.ByteString, string)",
            "name":"TestNetFee.TestNetFee,Update",
            "range":"476-513",
            "params":[
                "nefFile,ByteArray,0",
                "manifest,String,1"
            ],
            "return":"Void",
            "variables":[

            ],
            "sequence-points":[
                "479[0]96:17-96:27",
                "487[0]96:29-96:70",
                "507[0]97:13-97:64",
                "513[0]98:9-98:10"
            ]
        },
        {
            "id":"TestNetFee.TestNetFee.Destroy()",
            "name":"TestNetFee.TestNetFee,Destroy",
            "range":"518-549",
            "params":[

            ],
            "return":"Void",
            "variables":[

            ],
            "sequence-points":[
                "518[0]102:17-102:27",
                "526[0]102:29-102:70",
                "546[0]103:13-103:42",
                "549[0]104:9-104:10"
            ]
        },
        {
            "id":"TestNetFee.TestNetFee.DoRequest(string, long)",
            "name":"TestNetFee.TestNetFee,DoRequest",
            "range":"554-590",
            "params":[
                "url,String,0",
                "gasForResponse,Integer,1"
            ],
            "return":"Void",
            "variables":[
                "filter,String,0",
                "callback,String,1",
                "userdata,Any,2"
            ],
            "sequence-points":[
                "557[0]108:20-108:31",
                "560[0]109:20-109:41",
                "571[0]110:20-110:41",
                "582[0]111:13-111:77",
                "590[0]112:9-112:10"
            ]
        },
        {
            "id":"TestNetFee.TestNetFee.Callback(string, string, Neo.SmartContract.Framework.Native.OracleResponseCode, string)",
            "name":"TestNetFee.TestNetFee,Callback",
            "range":"595-752",
            "params":[
                "url,String,0",
                "userdata,String,1",
                "code,Integer,2",
                "result,String,3"
            ],
            "return":"Void",
            "variables":[
                "ret,Any,0",
                "arr,Array,1",
                "value,String,2"
            ],
            "sequence-points":[
                "598[0]116:17-116:57",
                "628[0]116:59-116:96",
                "644[0]117:17-117:51",
                "649[0]117:53-117:124",
                "693[0]119:20-119:56",
                "698[0]120:22-120:41",
                "700[0]121:20-121:42",
                "704[0]123:13-123:50",
                "725[0]124:13-124:53",
                "752[0]125:9-125:10"
            ]
        }
    ],
    "events":[
        {
            "id":"UnlockEvent",
            "name":"TestNetFee.TestNetFee,UnlockEvent",
            "params":[
                "arg1,Hash160,0",
                "arg2,Hash160,1",
                "arg3,Integer,2"
            ]
        },
        {
            "id":"IsPausedEvent",
            "name":"TestNetFee.TestNetFee,IsPausedEvent",
            "params":[
                "obj,Any,0"
            ]
        }
    ]
}
```

**Review by superboyiii — APPROVED** [review]

**Comment by erikzhang** [comment]
@superboyiii You didn't test instance methods?

**Comment by erikzhang** [comment]
@devhawk Does it work for you?

**Comment by superboyiii** [comment]
> @superboyiii You didn't test instance methods?

Instance method works as well.
```
{
            "id":"pnUSDT.pnUSDT.MintNFT(string, string, string)",
            "name":"pnUSDT.pnUSDT,MintNFT",
            "range":"1363-1377",
            "params":[
                "this,Any,0",
                "name,String,1",
                "description,String,2",
                "image,String,3"
            ],
            "return":"ByteArray",
            "variables":[

            ],
            "sequence-points":[
                "1366[0]165:13-165:61",
                "1377[0]166:9-166:10"
            ]
        }
    ]
```

**Comment by devhawk** [comment]
I'll give it a try tomorrow

**Review by shargon — APPROVED** [review]

**Comment by erikzhang** [comment]
Ping @devhawk

**Comment by devhawk** [comment]
Ill take a look in the morning my time

**Review by devhawk — APPROVED** [review]
It's no longer producing invalid debug info, but it appears the explicit indexes always match the implicit ones now

**Review by superboyiii — APPROVED** [review]

**Comment by erikzhang** [comment]
> but it appears the explicit indexes always match the implicit ones now

It's not good? @devhawk

**Comment by devhawk** [comment]
> It's not good? @devhawk

It's fine, but I don't understand why we needed to add explicit slot indexes to the debug into. The parameters array is just encoding the implicit array index explicitly. the explicit slot indexes *are* used in static-variables and variables array, but it would have been easier to leave the debug info format alone and put the slots for the "hidden" variables at the end after then ones that have associated names.

**Pull request merged**